### ISSUE TITLE
[BUG] Toggling manual controls on the Dashboard throws ControlNode thread warnings

### ISSUE THREAD

**Issue by cooperbell** [UIOWA5830SP19/SPP100#30]
**Describe the bug**
Toggling manual controls on the Dashboard throws ControlNode thread warnings.

**To Reproduce**
1. Run dashboard and boat code
2. Open dashboard
3. In the bottom left check the "Toggle Manual Controls" button

**Expected behavior**
Ideally, checking that box doesn't throw ControlNode thread warnings.

**Screenshots**
Terminal info:
```
[2019/01/28 10:00:13.119 AM] arch_w: ControlsNode has more than 5 active threads in parallel. Be careful to not excede 15
[2019/01/28 10:00:13.119 AM] arch_e: ControlsNode has more than 25 tasks running. Change how you handle the message results.
Exception in thread "pool-3-thread-21" java.lang.NullPointerException
	at com.imarc.marc_i.controls.ControlsNode.onReceiveMessage(ControlsNode.java:75)
	at com.imarc.marc_i.arch.nodes.Node.lambda$receiveMessage$0(Node.java:151)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
**Boat code or simulation code?**
Boat

**Windows or Mac?**
Mac (but probably both)

**Additional context**

**Comment by jwryan9** [comment]
```
[2019/02/24 09:11:41.389 PM] arch_w: ControlsNode has more than 5 active threads in parallel. Be careful to not exceed 15
[2019/02/24 09:11:41.389 PM] arch_e: ControlsNode has more than 25 tasks running. Change how you handle the message results.
Exception in thread "pool-3-thread-621" java.lang.NullPointerException
	at com.imarc.marc_i.controls.ControlsNode.onReceiveMessage(ControlsNode.java:75)
	at com.imarc.marc_i.arch.nodes.Node.lambda$receiveMessage$0(Node.java:151)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

It appears that toggling manual controls spawns an infinite amount of threads. Looking at "pool-3-thread-621" we got to 621 parallel threads before stopping the program

**Comment by cooperbell** [comment]
> ```
> [2019/02/24 09:11:41.389 PM] arch_w: ControlsNode has more than 5 active threads in parallel. Be careful to not exceed 15
> [2019/02/24 09:11:41.389 PM] arch_e: ControlsNode has more than 25 tasks running. Change how you handle the message results.
> Exception in thread "pool-3-thread-621" java.lang.NullPointerException
> 	at com.imarc.marc_i.controls.ControlsNode.onReceiveMessage(ControlsNode.java:75)
> 	at com.imarc.marc_i.arch.nodes.Node.lambda$receiveMessage$0(Node.java:151)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)
> ```
> It appears that toggling manual controls spawns an infinite amount of threads. Looking at "pool-3-thread-621" we got to 621 parallel threads before stopping the program

Maybe the error is in the dashboard code?

**Comment by jwryan9** [comment]
> > ```
> > [2019/02/24 09:11:41.389 PM] arch_w: ControlsNode has more than 5 active threads in parallel. Be careful to not exceed 15
> > [2019/02/24 09:11:41.389 PM] arch_e: ControlsNode has more than 25 tasks running. Change how you handle the message results.
> > Exception in thread "pool-3-thread-621" java.lang.NullPointerException
> > 	at com.imarc.marc_i.controls.ControlsNode.onReceiveMessage(ControlsNode.java:75)
> > 	at com.imarc.marc_i.arch.nodes.Node.lambda$receiveMessage$0(Node.java:151)
> > 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> > 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> > 	at java.lang.Thread.run(Thread.java:748)
> > ```
> > It appears that toggling manual controls spawns an infinite amount of threads. Looking at "pool-3-thread-621" we got to 621 parallel threads before stopping the program
> 
> 
> Maybe the error is in the dashboard code?

Could be causing receiveMessage method to be invoked an infinite number of times.

**Comment by jwryan9** [comment]
Estimating 2 points left on this.

**Comment by jwryan9** [comment]
1. Dashboard & BoatOS running
2. Toggle on manual controls on Dashboard
3. Sends above messages
4. Toggle off manual controls on Dashboard
5. Messages continue to be sent
6. Restart BoatOS, leave Dashboard running
7. Messages resume when BoatOS reboot is complete
8. Restart BoatOS and Dashboard
9. Messages stop until manual controls toggled

**Comment by jwryan9** [comment]
When running debugger in recieveMessage(): messages correctly report 'manual': true in data field when manual controls toggled in Dashboard. Correctly reports false as well.

**Comment by jwryan9** [comment]
> When running debugger in recieveMessage(): messages correctly report 'manual': true in data field when manual controls toggled in Dashboard. Correctly reports false as well.

Seems to alternate between sending true and false when toggled true on Dashboard.

**Comment by jwryan9** [comment]
> > When running debugger in recieveMessage(): messages correctly report 'manual': true in data field when manual controls toggled in Dashboard. Correctly reports false as well.
> 
> Seems to alternate between sending true and false when toggled true on Dashboard.
 Example message values.
```
JsonMessage { from: NetworkNode, to: ControlsNode, name: drive, data: {"lThrust":0.0,"rThrust":0.0,"manual":true} }
```

```
JsonMessage { from: MappingNode, to: ControlsNode, name: drive, data: {"x":1.6783934665731515,"y":-0.19828403079829607,"v":0.41197319814016153,"angle":0.0,"manual":false} }
```

**Comment by cooperbell** [comment]
> ```
> JsonMessage { from: NetworkNode, to: ControlsNode, name: drive, data: {"lThrust":0.0,"rThrust":0.0,"manual":true} }
> ```
> ```
> JsonMessage { from: MappingNode, to: ControlsNode, name: drive, data: {"x":1.6783934665731515,"y":-0.19828403079829607,"v":0.41197319814016153,"angle":0.0,"manual":false} }
> ```
This might be due to us not correctly suspending the speed test when we toggle manual controls. If you look into the `onSpeedTest()` method of MappingNode.java, manual controls is being set to false here:
```
vector.addProperty("manual", false);
```

So, the program seems to be flipping back and forth on manual controls. Not sure if this is related to the threading issue, just something that we've identified as wrong.

**Comment by aklinker1** [comment]
Ahh yes, large thread counts. I remember why I added those warnings now.

@cooperbell Is correct, every time `onReceiveMessage` is called, a new thread is spawned. However, `Exception in thread "pool-3-thread-621"` does not mean that there are 621 active threads. It means that that was the 621st. If you add a log statement [in this function](), you'll be able to get an accurate count. `Node#mMessageThreadCount` should stay below 5, that's what I arbitrarily chose as a good number per thread. Since there are potentially a lot of nodes, lets say 10 for a bit of an overstimate, I want to limit the number of `synchronized` calls I'm making. This is mostly because I don't have complete confidence in Java's multi threading capabilities. The `synchronized` keyword is a bad implementation of handling parallel work.

On the Dashboard, I believe it is set to send a manual control request every `requestAnimationFrame`, which is probably between `30-60` times per second, depending on the computer it is running on. Lets assume the 60 request/sec. This means each message has 17ms to complete the thread. You can time this by chaning this `executeRunnable` to 

```java
mMessageExecutor.executeRunnable(() -> {
    long start = System.currentTimeMillis();
    addToTaskCount(1);
    onReceiveMessage(message);
    addToTaskCount(-1);
    long duration = System.currentTimeMillis() - start;
    Log.v("arch", "Message completed in " + duration + "ms");
});
```

There is no reason a properly implemented case for `Node#onRecieveMessage` should take more than 1ms. If it is, you aren't implementing the `onRecieveMessage` message properly in whichever node is causing this issue.

To be clear, this could have been my fault, and a change I made at the competition lead to this error happening.

---

### Review of What __NOT__ to do for controlling the boat.

Do not use HTTPS requests (I.E., the `NetworkNode`) to send manual control commands. While at the competition the network was so packed with wifi signals, it was not viable.

If you want to actually control it the same way in the simulation as the physical boat, you need to send the requests to the `HardwareNode`. That node is responsible for interacting with the Arduino. The Arduino takes inputs from a RC receiver. If it is manual mode, it performs some smoothing of the RC's values and changes the propellers speeds directly without communicating with the NUC. If it is in autonomous mode, the Arduino ignores the signals from the RC receiver and uses serial communication to talk to the hardware node.

What this means is to properly simulate the fact that the manual controls aren't coming from the network node, you should not send messages to the `ControlsNode` from the `NetworkNode`. You should change the simulation's boat values directly (get the simulation using `Simulation.getSingleton()`) in the `ManualControlsHandler`. There probably shouldn't be any messages being passed around for this.

> ___PS, sorry for the essay lolz___

**Comment by jwryan9** [comment]
Closed by #132
